sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -556,6 +589,46 @@ private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignatur
     return DoFnSignature.SplitRestrictionMethod.create(m, restrictionT);
   }
 
+  private static ImmutableMap<String, TimerDeclaration> analyzeTimerDeclarations(
+      ErrorReporter errors, Class<?> fnClazz) {
+    Map<String, DoFnSignature.TimerDeclaration> declarations = new HashMap<>();
+    for (Field field : declaredFieldsWithAnnotation(DoFn.TimerId.class, fnClazz, DoFn.class)) {
+      String id = field.getAnnotation(DoFn.TimerId.class).value();
+
+      if (declarations.containsKey(id)) {
+        errors.throwIllegalArgument(
+            "Duplicate %s \"%s\", used on both of [%s] and [%s]",
+            DoFn.TimerId.class.getSimpleName(),
+            id,
+            field.toString(),
+            declarations.get(id).field().toString());
+        continue;
+      }
+
+      Class<?> timerSpecRawType = field.getType();
+      if (!(timerSpecRawType.equals(TimerSpec.class))) {
+        errors.throwIllegalArgument(
+            "%s annotation on non-%s field [%s] that has class %s",
```

Comment:
Yea, and actually it is fairly unreadable. I'll work refining the error messages, but certainly remove this redundancy.